README.md
# PyVer

<!-- markdownlint-disable MD013 -->
![Crates.io](https://img.shields.io/crates/l/pyver) ![Crates.io](https://img.shields.io/crates/v/pyver) ![docs.rs](https://img.shields.io/docsrs/pyver) [![🧪 Tests](https://github.com/Allstreamer/pyver/actions/workflows/tests.yml/badge.svg)](https://github.com/Allstreamer/pyver/actions/workflows/tests.yml) [![🖋  Check linting](https://github.com/Allstreamer/pyver/actions/workflows/lint.yml/badge.svg)](https://github.com/Allstreamer/pyver/actions/workflows/lint.yml) [![🔨 Build](https://github.com/Allstreamer/pyver/actions/workflows/build.yml/badge.svg)](https://github.com/Allstreamer/pyver/actions/workflows/build.yml) [![📦 Package](https://github.com/Allstreamer/pyver/actions/workflows/package.yml/badge.svg)](https://github.com/Allstreamer/pyver/actions/workflows/package.yml) [![📄 Build docs](https://github.com/Allstreamer/pyver/actions/workflows/docs.yml/badge.svg)](https://github.com/Allstreamer/pyver/actions/workflows/docs.yml) [![👔 Check formatting](https://github.com/Allstreamer/pyver/actions/workflows/format.yml/badge.svg)](https://github.com/Allstreamer/pyver/actions/workflows/format.yml)
<!-- markdownlint-enable MD013 -->

> **Python PEP-440 Version Parser**

This package allows for parsing Python [PEP-440](https://peps.python.org/pep-0440/)
version numbers and for comparisons between
[PEP-440](https://peps.python.org/pep-0440/) version numbers.

## Usage

```toml
[dependencies]
pyver = "1"
```

The following is an example for initilizing and comparing two version strings

```rust
use pyver::PackageVersion;

let a = PackageVersion::new("v1.0a2.dev456").unwrap();
let b = PackageVersion::new("v1.0a2.dev457").unwrap();

assert_eq!(a < b, true);
```

Comparing single version components

```rust
use pyver::PackageVersion;

let a = PackageVersion::new("1!1.323.dev2").unwrap();
let b = PackageVersion::new("v3.2.dev2").unwrap();

// Check that both have the same dev version
assert_eq!(a.dev, b.dev);
```

Seperation of version identifiers

```rust
use pyver::PackageVersion;

let version = PackageVersion::new("v1.23.dev2").unwrap();

println!("{:?}", version.release.major);
// > 1

println!("{:?}", version.release.minor);
// > 2

println!("{:?}", version.dev);
// > Some(DevHead { dev_num: Some(2) })
```

See more examples at the [docs](https://docs.rs/pyver/latest/pyver/)

## Contribution

For now Contributions will be quite loose.


CHANGELOG.md
<!--Markdownlint Rules-->
<!-- markdownlint-disable no-duplicate-header-->

# Changelog

<!-- All notable changes to this project will be documented in this file.
The format is based on :

* [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to:
* [Semantic Versioning](https://semver.org/spec/v2.0.0.html). -->

## Unreleased

### Added

* Added Github Actions for (cargo check, cargo fmt, cargo clippy, cargo test) [#10](https://github.com/Allstreamer/pyver/pull/10)
* Added CHANGELOG.md
* Added `ids` module
* Added the capability to use the `PackageVersion` struct as a HashMap key [#28](https://github.com/Rust-Python-Packaging/pyver/pull/28)

### Changed

## [1.0.1] - 2022-09-06

* [MINOR] Fixed Readme for crates.io

## [1.0.0] - 2022-09-06

* Added LICENSE with MIT License [#10](https://github.com/Allstreamer/pyver/pull/10)
* Added CI Badges to Readme by [#10](https://github.com/Allstreamer/pyver/pull/10)
* Moved PackageVersion & it's tests to version.rs [#21](https://github.com/Allstreamer/pyver/pull/21)
* Moved Identifiers & tests to `ids` module

### Docs

* Added Struct descriptions to each struct [#15](https://github.com/Allstreamer/pyver/pull/15)
* Added Examples to Rust Docs [#15](https://github.com/Allstreamer/pyver/pull/15)

## [0.1.0] - 2022-09-03

### Added

* Created Baseline Project


src/lib.rs
/*!
# Handling of `PEP-440`
This library implements Python's Package versioning system.

Read more at <https://peps.python.org/pep-0440/>

# Usage
The `pyver` crate is available on [crates.io](https://crates.io/crates/pyver),
you can include it in your project by adding the following to your `Cargo.toml`.
```toml
[dependencies]
pyver = "1.0"
```
# Example
The following example shows how to parse a package version and
how to compare them
```
use pyver::PackageVersion;

let a = PackageVersion::new("v1.0a2.dev456").unwrap();
let b = PackageVersion::new("v1.1a2.dev457").unwrap();

assert!(a < b);
```

If you want to verify single version strings do
```
use pyver::validate_440_version;

assert!(
    validate_440_version("1.0").is_ok()
);
```
*/

#[macro_use]
extern crate derivative;

mod validator;
// Expose validate_440_version function
pub use validator::validate_440_version;

/// Identifiers (i.e. the components of a version string)
// Expose Ids Module
pub mod ids;

mod version;
// Expose PackageVersion Struct
pub use version::PackageVersion;


src/validator.rs
use anyhow::Result;
use lazy_static::lazy_static;
use pomsky_macro::pomsky;
use regex::Captures;

/// Utility Function for Checking if a `PEP-440` Version String is valid
/// and getting it's groups
///
/// # Example Usage
/// ```
/// use pyver::validate_440_version;
///
/// match validate_440_version("v1.0") {
///     Ok(v) => println!("Is a valid Version!"),
///     Err(e) => panic!("Not Valid!"),
/// }
/// ```
pub fn validate_440_version(version: &str) -> Result<Captures> {
    lazy_static! {
        // Safe to unwrap since Regex is predefined
        // Regex as defined in PEP-0440
        static ref VERSION_VALIDATOR: regex::Regex =
            regex::Regex::new(VALIDATION_REGEX).unwrap();
    }

    // Capture each group of the regex
    // Groups are:
    // epoch, release, pre, pre_l, pre_n, post, post_l, post_n1, post_n2,
    // dev, dev_l, dev_n, local
    let version_match: Captures = match VERSION_VALIDATOR.captures(version) {
        Some(v) => v,
        None => anyhow::bail!("Failed to decode version {}", version),
    };
    Ok(version_match)
}

/// Rulex version of
/// Python's PEP-440 Regex
/// (<https://peps.python.org/pep-0440/#appendix-b-parsing-version-strings-with-regular-expressions>)
static VALIDATION_REGEX: &str = pomsky!(
    // Version String may start with v<version_number>
    // Example:
    // v1.0
    "v"?

    // Version String may include an epoch <epoch_num>!<version>
    // Example:
    // 1!1.0
    (:epoch(['0'-'9']+)'!')?

    // Version String must include major and minor version <major>.<minor>
    // Example:
    // 1.0
    :release(['0'-'9']+("."['0'-'9']+)*)

    // Version String may include Pre-Header
    // Example:
    // 1.0.preview-2
    // 1.0.rc2
    // 1.0beta2
    :pre(
        ["-" "_" "."]?

        :pre_l(
        ("preview"|"alpha"|"beta"|"pre"|"rc"|"a"|"b"|"c")
        )

        ["-" "_" "."]?

        :pre_n(['0'-'9']+)?
    )?

    // Version String may include Post-Header
    // Examples:
    // 1.0-9
    // 1.0-post2
    // 1.0.post.2
    :post(
        "-"
        :post_n1(['0'-'9']+)

        |

        ["-" "_" "."]?
        :post_l("post" | "rev" | "r")
        ["-" "_" "."]?
        :post_n2(['0'-'9']+)?
    )?

    // Version string may include Dev-header
    // Example:
    // 1.0-dev3
    // 1.0dev4
    // 1.0_dev_9
    :dev(
        ["-" "_" "."]?
        :dev_l("dev")
        ["-" "_" "."]?
        :dev_n(['0'-'9']+)?
    )?

    // Version string may include Local Version
    // Local version must start with +
    // Example:
    // 1.0+this.can.say.anything.as.long.as.its.a.letter.or.number.231241
    (
    "+"
    :local(
        ['a'-'z' '0'-'9']+
        ((["-" "_" "."] ['a'-'z' '0'-'9']+)+)?
    )
    )?
);


src/ids/dev_id.rs
use serde::{Deserialize, Serialize};

/// # `PEP-440` Developmental release identifier
/// This identifier is used to mark a developmental release
///
/// Examples of versions that use this struct:
/// - `1.0.dev456`
/// - `1.0rc1.dev1`
///
/// ## Example Usage
/// ```
/// use pyver::ids::DevHead;
///
/// assert!(
///     DevHead { dev_num: Some(0) }
///     >
///     DevHead { dev_num: None }
/// );
/// ```
#[derive(
    Hash, Ord, Clone, Debug, Serialize, Deserialize, Eq, PartialEq, PartialOrd,
)]
pub struct DevHead {
    pub dev_num: Option<u32>,
}

#[cfg(test)]
mod tests {
    use super::DevHead;

    #[test]
    fn test_dev_ordering() {
        assert!(DevHead { dev_num: Some(0) } > DevHead { dev_num: None });
        assert!(DevHead { dev_num: Some(1) } > DevHead { dev_num: Some(0) });
    }
}


src/ids/pre_id.rs
use serde::{Deserialize, Serialize};

/// # `PEP-440` Pre-Release identifier
/// This identifier is used to mark a Pre-Release version
#[derive(
    Hash, Ord, Clone, Debug, Serialize, Deserialize, Eq, PartialEq, PartialOrd,
)]
pub enum PreHeader {
    /// Present in versions like 1.1beta1 or 1.0b1 both are represented the same way
    /// ```
    /// use pyver::ids::PreHeader;
    ///
    /// PreHeader::Beta(Some(1));
    /// ```
    Beta(Option<u32>),
    /// Present in versions like 1.0alpha2 or 1.0a2 both are represented the same way
    /// ```
    /// use pyver::ids::PreHeader;
    ///
    /// PreHeader::Alpha(Some(2));
    /// ```
    Alpha(Option<u32>),
    /// Present in versions like 1.1pre3
    /// ```
    /// use pyver::ids::PreHeader;
    ///
    /// PreHeader::Preview(Some(3));
    /// ```
    Preview(Option<u32>),
    /// Present in versions like 1.1-rc-4 or 1.1c-4
    /// ```
    /// use pyver::ids::PreHeader;
    ///
    /// PreHeader::ReleaseCandidate(Some(4));
    /// ```
    ReleaseCandidate(Option<u32>),
}

#[cfg(test)]
mod tests {
    use super::PreHeader;

    #[test]
    fn test_pre_ordering() {
        assert!(PreHeader::ReleaseCandidate(None) > PreHeader::Preview(None));
        assert!(PreHeader::Preview(None) > PreHeader::Alpha(None));
        assert!(PreHeader::Alpha(None) > PreHeader::Beta(None));

        assert!(
            PreHeader::ReleaseCandidate(Some(2)) > PreHeader::ReleaseCandidate(Some(1))
        );
        assert!(PreHeader::Preview(Some(50)) > PreHeader::Preview(Some(3)));
        assert!(PreHeader::Alpha(Some(504)) > PreHeader::Alpha(Some(0)));
        assert!(PreHeader::Beta(Some(1234)) > PreHeader::Beta(Some(1)));

        assert!(PreHeader::ReleaseCandidate(Some(1)) > PreHeader::Beta(Some(45067885)));
    }
}


src/ids/post_id.rs
use serde::{Deserialize, Serialize};
use std::cmp::Ordering;

/// # `PEP-440` Post-Release identifier
/// This identifier is used to mark a Post Release/Revision Version
///
/// Examples of versions that use this struct:
/// - `1.0.post456`
/// - `1.0rev`
///
/// ## Example Usage
/// ```
/// use pyver::ids::PostHeader;
/// use pyver::ids::PostHead;
///
/// assert!(
///     PostHeader {
///         post_head: Some(PostHead::Post),
///         post_num: Some(0),
///     } > PostHeader {
///         post_head: Some(PostHead::Post),
///         post_num: None,
///     }
/// );
/// ```
#[derive(Hash, Clone, Debug, Serialize, Deserialize, Eq, PartialEq)]
pub struct PostHeader {
    pub post_head: Option<PostHead>,
    pub post_num: Option<u32>,
}

/// `PEP-440` Post-Release Identifier Keyword
/// This is a helper enum to tack whether it's a Revision or
/// a Post-Release
///
/// Examples of versions that use this enum:
/// - `1.0.post456`
/// - `1.0rev`
#[derive(Hash, Clone, Debug, Serialize, Deserialize, Eq, PartialEq)]
pub enum PostHead {
    /// ```
    /// use pyver::ids::PostHead;
    ///
    /// PostHead::Post;
    /// ```
    Post,
    /// ```
    /// use pyver::ids::PostHead;
    ///
    /// PostHead::Rev;
    /// ```
    Rev,
}

impl Ord for PostHead {
    fn cmp(&self, _other: &Self) -> Ordering {
        Ordering::Equal
    }
}

impl PartialOrd for PostHead {
    fn partial_cmp(&self, _other: &Self) -> Option<Ordering> {
        Some(Ordering::Equal)
    }
}

impl PartialOrd<Self> for PostHeader {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        Some(self.cmp(other))
    }
}

impl Ord for PostHeader {
    fn cmp(&self, other: &Self) -> Ordering {
        if self.post_num == other.post_num {
            return Ordering::Equal;
        }

        if self.post_num.is_none() && other.post_num.is_some() {
            return Ordering::Less;
        } else if self.post_num.is_some() && other.post_num.is_none() {
            return Ordering::Greater;
        }

        if self.post_num < other.post_num {
            Ordering::Less
        } else {
            Ordering::Greater
        }
    }
}

#[cfg(test)]
mod tests {
    use super::PostHead;
    use super::PostHeader;

    #[test]
    fn test_post_ordering() {
        assert!(
            PostHeader {
                post_head: Some(PostHead::Post),
                post_num: Some(0),
            } > PostHeader {
                post_head: Some(PostHead::Post),
                post_num: None,
            }
        );
        assert!(
            PostHeader {
                post_head: Some(PostHead::Post),
                post_num: Some(1),
            } > PostHeader {
                post_head: Some(PostHead::Post),
                post_num: Some(0),
            }
        );
    }
}


src/ids/release_id.rs
use serde::{Deserialize, Serialize};

/// `PEP-440` Release numbers
#[derive(
    Hash, Ord, Clone, Debug, Serialize, Deserialize, Eq, PartialEq, PartialOrd,
)]
pub struct ReleaseHeader {
    /// Major release such as 1.0 or breaking changes
    pub major: u32,
    /// Minor release Such as new functionality
    pub minor: u32,
}

#[cfg(test)]
mod test {
    use super::ReleaseHeader;

    #[test]
    fn test_release_ordering() {
        assert!(
            ReleaseHeader { major: 1, minor: 0 } > ReleaseHeader { major: 0, minor: 0 }
        );
        assert!(
            ReleaseHeader { major: 1, minor: 1 } > ReleaseHeader { major: 1, minor: 0 }
        );
        assert!(
            ReleaseHeader { major: 2, minor: 1 }
                > ReleaseHeader {
                    major: 1,
                    minor: 52,
                }
        );
    }
}


src/version.rs
use super::ids::{DevHead, PostHead, PostHeader, PreHeader, ReleaseHeader};
use super::validate_440_version;
use anyhow::Result;
use serde::{Deserialize, Serialize};
use std::fmt;
use std::hash::{Hash, Hasher};

/// `PEP-440` Compliant versioning system
///
//# This struct is sorted so that PartialOrd
//# correctly interprets priority
//# Lower == More important
///
/// # Example Usage
/// ```
///# use pyver::PackageVersion;
/// let _ = PackageVersion::new("v1.0");
/// ```
#[derive(Clone, Derivative, Debug, Serialize, Deserialize)]
#[derivative(PartialOrd, Ord)]
pub struct PackageVersion {
    /// ## Original String
    /// Just holds the original string passed in when creating
    /// the `PackageVersion` as some formating data is lost
    /// when parsing the string
    #[derivative(PartialOrd = "ignore", PartialEq = "ignore")]
    pub original: String,

    /// ## `PEP-440` Local version identifier
    /// Local version sorting will have to be it's own issue
    /// since there are no limits to what a local version can be
    ///
    /// For those who can read regex here it is for the local version:
    /// `[a-z0-9]+(?:(?:[\-_.][a-z0-9]+)+)?`
    ///
    /// Here in Rulex:
    /// ```toml
    ///  ['a'-'z' '0'-'9']+
    ///  ((["-" "_" "."] ['a'-'z' '0'-'9']+)+)?
    /// ```
    #[derivative(PartialOrd = "ignore", PartialEq = "ignore")]
    pub local: Option<String>,

    /// ## `PEP-440` Developmental release identifier
    pub dev: Option<DevHead>,

    /// ## `PEP-440` Post-Release identifier
    pub post: Option<PostHeader>,

    /// ## `PEP-440` Pre-Release identifier
    pub pre: Option<PreHeader>,

    /// ## `PEP-440` Release number
    pub release: ReleaseHeader,

    /// ## `PEP-440` Version-Epoch
    pub epoch: Option<u32>,
}

impl PackageVersion {
    pub fn new(version: &str) -> Result<Self> {
        let version_match = validate_440_version(version)?;

        let epoch: Option<u32> = match version_match.name("epoch") {
            // Convert Epoch String to Epoch Number
            Some(v) => Some(v.as_str().parse::<u32>()?),
            None => None,
        };

        let release: ReleaseHeader = match version_match.name("release") {
            Some(v) => {
                // Does Release String contain minor version
                if v.as_str().contains('.') {
                    let split: Vec<&str> = v.as_str().split('.').into_iter().collect();
                    ReleaseHeader {
                        major: split[0].parse::<u32>()?,
                        minor: split[1].parse::<u32>()?,
                    }
                } else {
                    ReleaseHeader {
                        major: v.as_str().parse::<u32>()?,
                        minor: 0,
                    }
                }
            }
            // There always has to be at least a major version
            None => anyhow::bail!("Failed to decode version {}", version),
        };

        let pre: Option<PreHeader> = match version_match.name("pre") {
            Some(_) => {
                let pre_n = match version_match.name("pre_n") {
                    Some(v) => Some(v.as_str().parse::<u32>()?),
                    None => None,
                };

                // Should be safe to unwrap since we already checked if pre has a value
                // since pre_n has to exist
                match version_match.name("pre_l").unwrap().as_str() {
                    "alpha" => Some(PreHeader::Alpha(pre_n)),
                    "a" => Some(PreHeader::Alpha(pre_n)),
                    "beta" => Some(PreHeader::Beta(pre_n)),
                    "b" => Some(PreHeader::Beta(pre_n)),
                    "rc" => Some(PreHeader::ReleaseCandidate(pre_n)),
                    "c" => Some(PreHeader::ReleaseCandidate(pre_n)),
                    "preview" => Some(PreHeader::Preview(pre_n)),
                    "pre" => Some(PreHeader::Preview(pre_n)),
                    _ => None,
                }
            }
            None => None,
        };

        let post: Option<PostHeader> = match version_match.name("post") {
            Some(_) => {
                let post_num: Option<u32> = match version_match.name("post_n1") {
                    Some(v) => Some(v.as_str().parse::<u32>()?),
                    None => match version_match.name("post_n2") {
                        Some(v) => Some(v.as_str().parse::<u32>()?),
                        _ => None,
                    },
                };

                let post_head: Option<PostHead> = match version_match.name("post_l") {
                    Some(v) => {
                        match v.as_str() {
                            "post" => Some(PostHead::Post),
                            "rev" => Some(PostHead::Rev),
                            "r" => Some(PostHead::Rev),
                            // This branch Should be impossible (see regex-group post_l)
                            _ => None,
                        }
                    }
                    None => None,
                };

                Some(PostHeader {
                    post_head,
                    post_num,
                })
            }
            None => None,
        };

        let dev: Option<DevHead> = match version_match.name("dev") {
            Some(_) => {
                let dev_num = match version_match.name("dev_n") {
                    Some(v) => Some(v.as_str().parse::<u32>()?),
                    None => None,
                };
                Some(DevHead { dev_num })
            }
            None => None,
        };

        let local: Option<String> =
            version_match.name("local").map(|v| v.as_str().to_string());

        Ok(Self {
            original: version.to_string(),
            epoch,
            release,
            pre,
            post,
            dev,
            local,
        })
    }
}

impl fmt::Display for PackageVersion {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(f, "{}", self.original)
    }
}

impl PartialEq<Self> for PackageVersion {
    fn eq(&self, other: &Self) -> bool {
        self.release == other.release
            && self.local == other.local
            && self.dev == other.dev
            && self.epoch == other.epoch
            && self.post == other.post
            && self.pre == other.pre
    }
}

impl Eq for PackageVersion {}

/// The hash of the `PackageVersion` is calculated only from the `original` field.
impl Hash for PackageVersion {
    fn hash<H: Hasher>(&self, state: &mut H) {
        self.release.hash(state);
        self.local.hash(state);
        self.dev.hash(state);
        self.epoch.hash(state);
        self.post.hash(state);
        self.pre.hash(state);
    }
}

#[cfg(test)]
mod tests {
    use crate::PackageVersion;
    use anyhow::Result;
    use std::collections::hash_map::DefaultHasher;
    use std::collections::HashMap;
    use std::hash::{Hash, Hasher};

    fn default_hash<T: Hash>(value: &T) -> u64 {
        let mut hasher = DefaultHasher::new();
        value.hash(&mut hasher);
        hasher.finish()
    }

    #[test]
    fn test_pep440_ordering() -> Result<()> {
        assert!(
            PackageVersion::new(
                "v1!1.0-preview-921.post-516.dev-241+yeah.this.is.the.problem.with.local.versions",
            )?
                >
                PackageVersion::new("1.0")?
        );
        Ok(())
    }

    #[test]
    fn test_pep440_equality() -> Result<()> {
        assert_eq!(
            PackageVersion::new("1.0a1")?,
            PackageVersion::new("1.0alpha1")?
        );
        assert_eq!(
            PackageVersion::new("1.0b")?,
            PackageVersion::new("1.0beta")?
        );
        assert_eq!(PackageVersion::new("1.0r")?, PackageVersion::new("1.0rev")?);
        assert_eq!(PackageVersion::new("1.0c")?, PackageVersion::new("1.0rc")?);
        assert_eq!(PackageVersion::new("v1.0")?, PackageVersion::new("1.0")?);
        Ok(())
    }

    #[test]
    fn test_pep440() {
        // list of every example mentioned in pep-440
        let versions = vec![
            "1.0",
            "v1.1",
            "2.0",
            "2013.10",
            "2014.04",
            "1!1.0",
            "1!1.1",
            "1!2.0",
            "2!1.0.pre0",
            "1.0.dev456",
            "1.0a1",
            "1.0a2.dev456",
            "1.0a12.dev456",
            "1.0a12",
            "1.0b1.dev456",
            "1.0b2",
            "1.0b2.post345.dev456",
            "1.0b2.post345",
            "1.0rc1.dev456",
            "1.0rc1",
            "1.0",
            "1.0+abc.5",
            "1.0+abc.7",
            "1.0+5",
            "1.0.post456.dev34",
            "1.0.post456",
            "1.0.15",
            "1.1.dev1",
        ];

        for version in versions {
            match PackageVersion::new(version) {
                Ok(_v) => continue,
                Err(e) => panic!("Oh no {}", e),
            }
        }
    }

    #[test]
    fn test_pep440_negative() {
        let versions = vec!["not a version"];

        for version in versions {
            match PackageVersion::new(version) {
                Ok(v) => panic!("Oh no {}", v),
                Err(_e) => continue,
            }
        }
    }

    #[test]
    fn test_use_package_version_as_hash_key() -> Result<()> {
        let versions = vec![
            "1.0",
            "v1.1",
            "2.0",
            "2013.10",
            "2014.04",
            "1!1.0",
            "1!1.1",
            "1!2.0",
            "2!1.0.pre0",
            "1.0.dev456",
            "1.0a1",
            "1.0a2.dev456",
            "1.0a12.dev456",
            "1.0a12",
            "1.0b1.dev456",
            "1.0b2",
            "1.0b2.post345.dev456",
            "1.0b2.post345",
            "1.0rc1.dev456",
            "1.0rc1",
            "1.0",
            "1.0+abc.5",
            "1.0+abc.7",
            "1.0+5",
            "1.0.post456.dev34",
            "1.0.post456",
            "1.0.15",
            "1.1.dev1",
        ];

        let mut some_hash: HashMap<PackageVersion, String> = HashMap::new();

        for i in versions {
            some_hash.insert(PackageVersion::new(i).unwrap(), i.to_string());
        }

        some_hash.get(&PackageVersion::new("1.0").unwrap());

        Ok(())
    }

    #[test]
    fn test_hashing_and_eq() -> Result<()> {
        assert_eq!(
            default_hash(&PackageVersion::new("1.0a1")?),
            default_hash(&PackageVersion::new("1.0alpha1")?)
        );
        assert_eq!(
            default_hash(&PackageVersion::new("1.0b")?),
            default_hash(&PackageVersion::new("1.0beta")?)
        );
        assert_eq!(
            default_hash(&PackageVersion::new("1.0r")?),
            default_hash(&PackageVersion::new("1.0rev")?)
        );
        assert_eq!(
            default_hash(&PackageVersion::new("1.0c")?),
            default_hash(&PackageVersion::new("1.0rc")?)
        );
        assert_eq!(
            default_hash(&PackageVersion::new("v1.0")?),
            default_hash(&PackageVersion::new("1.0")?)
        );

        Ok(())
    }
}
